modules/aggregated-dex/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+
+		if let Some(swap_result) =
+			T::StableAsset::get_swap_output_amount(pool_id, input_index, output_index, supply_amount)
+		{
+			if swap_result.dy >= min_target_amount {
+				return Some((swap_result.dx, swap_result.dy));
+			}
+		}
+
+		None
+	}
+
+	#[transactional]
+	fn swap(
+		who: &T::AccountId,
+		supply_currency_id: CurrencyId,
+		target_currency_id: CurrencyId,
+		limit: SwapLimit<Balance>,
+	) -> sp_std::result::Result<(Balance, Balance), DispatchError> {
+		let (supply_amount, min_target_amount) = match limit {
+			SwapLimit::ExactSupply(supply_amount, min_target_amount) => (supply_amount, min_target_amount),
+			SwapLimit::ExactTarget(_, _) => return Err(Error::<T>::CannotSwap.into()),
+		};
+
+		let (pool_id, input_index, output_index, _) =
+			T::StableAsset::get_best_route(supply_currency_id, target_currency_id, min_target_amount)
+				.ok_or(Error::<T>::CannotSwap)?;
+		let pool_info = T::StableAsset::pool(pool_id).ok_or(Error::<T>::InvalidPoolId)?;
+		let asset_length = pool_info.assets.len() as u32;
+
+		let (actual_supply, actual_target) = T::StableAsset::swap(
+			who,
+			pool_id,
+			input_index,
+			output_index,
+			supply_amount,
+			min_target_amount,
+			asset_length,
+		)?;
+		ensure!(actual_target >= min_target_amount, Error::<T>::CannotSwap);
+
+		Ok((actual_supply, actual_target))
+	}
+}
+
+/// Choose DEX or Taiga to fully execute the swap by which price is better.
+pub struct EitherDexOrTaigaSwap<T>(PhantomData<T>);
+impl<T: Config> Swap<T::AccountId, Balance, CurrencyId> for EitherDexOrTaigaSwap<T> {
+	fn get_swap_amount(
+		supply_currency_id: CurrencyId,
+		target_currency_id: CurrencyId,
+		limit: SwapLimit<Balance>,
+	) -> Option<(Balance, Balance)> {
+		let dex_result = DexSwap::<T>::get_swap_amount(supply_currency_id, target_currency_id, limit);
+		let taiga_result = TaigaSwap::<T>::get_swap_amount(supply_currency_id, target_currency_id, limit);
+
+		if let (Some((dex_supply, dex_target)), Some((taiga_supply, taiga_target))) = (dex_result, taiga_result) {
+			if dex_supply > taiga_supply || dex_target < taiga_target {
+				taiga_result
+			} else {
+				dex_result
+			}
+		} else {
+			dex_result.or(taiga_result)
+		}
+	}
+
+	fn swap(
+		who: &T::AccountId,
+		supply_currency_id: CurrencyId,
+		target_currency_id: CurrencyId,
+		limit: SwapLimit<Balance>,
+	) -> sp_std::result::Result<(Balance, Balance), DispatchError> {
+		let dex_result = DexSwap::<T>::get_swap_amount(supply_currency_id, target_currency_id, limit);
+		let taiga_result = TaigaSwap::<T>::get_swap_amount(supply_currency_id, target_currency_id, limit);
+
+		let do_swap =
+			if let (Some((dex_supply, dex_target)), Some((taiga_supply, taiga_target))) = (dex_result, taiga_result) {
+				if dex_supply > taiga_supply || dex_target < taiga_target {
+					taiga_result
+				} else {
+					dex_result
+				}
+			} else {
+				dex_result.or(taiga_result)
+			};
```

Comment:
adding a `get_swap_params` method to `EitherDexOrTaigaSwap` would help to merge these duplication with `get_swap_amount` impl, like:

```rust
struct DexOrTaigaSwapParams {
    dex_result: ...,
    taiga_result: ...,
    swap_amount: Option<(Balance, Balance),
}

impl<T> for EitherDexOrTaigaSwap<T> {
    fn get_swap_params(
		supply_currency_id: CurrencyId,
		target_currency_id: CurrencyId,
		limit: SwapLimit<Balance>,
	) -> DexOrTaigaSwapParams {
	    // ...
	}
}
```